### PR TITLE
Data Parallel Training Workload for LLaMA 3.2 1B - LoRA Finetuning 

### PR DESCRIPTION
**Pull request by pglusacTT**
### Ticket
N/A

### Model training setting

- **Model:** `LLaMA 3.2 1B`
- **Strategy:** SFT with LoRA layers on a T3K
- **Framework:** PyTorch
- **Task:** Sentiment prediction using `SST2`

### What's changed

Implement data parallel training workload for LLaMA 3.2 1B

### Experiment results
image

### Checklist
- [x] Main training workflow
- [x] W&B logging support
- [ ] CPU testing
- [ ] GPU testing
- [x] TT device testing

### CODE REVIEW

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 1]
```diff
@@ -117,6 +156,20 @@ def train(config: TrainingConfig, device: torch.device, logger: TrainingLogger,
...
+                # TODO: Refactor when https://github.com/tenstorrent/tt-xla/issues/1993 is resolved.
+                if config.parallelism == "data":
+                    expected_output, labels_mask = transform_labels(
+                        batch, config.ignored_index, model.model.config.vocab_size, device
+                    )
+
+                    # Apply sharding on inputs.
+                    import torch_xla.distributed.spmd as xs
+
+                    xs.mark_sharding(input_ids, mesh, ("data", None))
```
Comment (umalesTT) [inline blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: For my own curiosity, how do we make sure parameters of model are replicated? 
I don't see we did that anywhere explicitly

**Review by umalesTT — COMMENTED** [review]
Really great

File: blacksmith/tools/torch_xla_utils.py [hunk 2]
```diff
@@ -34,3 +34,32 @@ def init_device(plugin_path: str = DEFAULT_PJRT_PATH):
     plugin = TTPjrtPlugin(path)
```
Comment (pmarkovicTT) [inline blacksmith/tools/torch_xla_utils.py]: General question, we don't need `TTPjrtPlugin` and `init_device`, right? @umalesTT 
If so, can you clean them from this script, along with 
```
os.environ["PJRT_DEVICE"] = "TT"
os.environ["XLA_STABLEHLO_COMPILE"] = "1"
DEFAULT_PJRT_PATH = "third_party/tt-xla/build/src/tt/pjrt_plugin_tt.so"
```

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 3]
```diff
@@ -7,18 +7,48 @@
...
 from blacksmith.tools.cli import generate_config
 from blacksmith.tools.torch_helpers import show_examples, collect_examples, collate_fn_for_causal_lm
+from blacksmith.tools.torch_xla_utils import setup_multi_chip_environment, get_mesh
 from blacksmith.tools.logging_manager import TrainingLogger
 from blacksmith.tools.checkpoints_manager import CheckpointManager
 from blacksmith.tools.reproducibility_manager import ReproducibilityManager
 
 
+# Custom cross-entropy loss because of https://github.com/tenstorrent/tt-xla/issues/1993.
+def cross_entropy_loss(shift_logits, expected_output, labels_mask):
```
Comment (pmarkovicTT) [inline blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: `cross_entropy_loss` and `transform_labels` should be put in a separate utils file.

**Review by pmarkovicTT — COMMENTED** [review]
great work, just make sure to align with other multi-chip setups in the repo

File: blacksmith/tools/torch_xla_utils.py (hunk 2 above)
Comment (umalesTT) [reply blacksmith/tools/torch_xla_utils.py]: I don't think we need this, although I can't claim this for torch workloads 100%
For Jax workloads this is the case

**Review by mmilosevicTT — COMMENTED** [review]
Great work, left some nit comments 😄

File: blacksmith/tools/torch_xla_utils.py [hunk 4]
```diff
@@ -1,65 +1,45 @@
...
+    mesh_shape = None
+    axis_names = None
 
-    if config.parallelism != "single":
-        num_devices = xr.global_runtime_device_count()
+    if config.parallelism == "data":
         mesh_shape = (num_devices, 1)
-        device_ids = np.array(range(num_devices))
-        mesh = Mesh(device_ids, mesh_shape, ("data", "model"))
+        axis_names = ("data", "model")
```
Comment (umalesTT) [inline blacksmith/tools/torch_xla_utils.py]: We don't need "model" axis here, right?

Comment (umalesTT) [reply blacksmith/tools/torch_xla_utils.py]: axis_names provided here will be used for dp+tp workloads

File: blacksmith/experiments/torch/llama/xla/test_dp_llama_fine_tuning_pure_torch.yaml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+# LoRA setup
+lora_r: 4
+lora_alpha: 8
+lora_target_modules: ["q_proj", "v_proj"]
+lora_task_type: "CAUSAL_LM"
+
+# Multi-chip settings
+parallelism: "data"  # [single, data, tensor]
+mesh_shape: "8,1"  # only used if parallelism != "single"
```
[the remark this answers is not in the document]
Comment (umalesTT) [reply blacksmith/experiments/torch/llama/xla/test_dp_llama_fine_tuning_pure_torch.yaml]: In README too

**Review by umalesTT — APPROVED** [review]

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 6]
```diff
@@ -174,9 +205,20 @@ def train(config: TrainingConfig, device: torch.device, logger: TrainingLogger,
 
 if __name__ == "__main__":
     # Config setup
-    config_file_path = os.path.join(os.path.dirname(__file__), "test_llama_fine_tuning_pure_torch.yaml")
+    parser = argparse.ArgumentParser(description="LLaMA Fine-Tuning with PyTorch and XLA")
+    parser.add_argument("--config", type=str, required=False, help="Path to the configuration YAML file.")
```
Comment (pmarkovicTT) [inline blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: I noted down this for discussion (we need to define/agree on work with multiple configs/setups)

**Review by pmarkovicTT — APPROVED** [review]
lgtm, align with @abogdanovicTT to merge
